src/parser.ts

```diff
@@ -37,48 +44,90 @@ const stateSet = {
     [TOKEN_TYPES.closeBlock]: { match: ']', push: 'main' },
   },
 };
-export function parser(config: string) {
+export function tokenize(config: string, filePath: string, pathMemo = new Set()): FluentBitToken[] {
   if (!config.replace(/\s/g, '')) {
-    throw new Error('Invalid config file');
+    throw new TokenError('File is empty', filePath, 0, 0);
   }
 
   if (!isFluentBit(config)) {
-    throw new Error('This file is not a valid Fluent Bit config file');
+    throw new TokenError('This file is not a valid Fluent Bit config file', filePath, 0, 0);
   }
 
-  const lexer = states(stateSet);
-  const tokens = lexer.reset(config);
+  let tokens = [] as FluentBitToken[];
+  const lexer = states(stateSet).reset(config);
+
+  // We will expand every include first, looking for any missing paths and invalid tokens
+  // https://github.com/calyptia/fluent-bit-config-parser/issues/15
+  for (const token of lexer) {
+    if (token.type === TOKEN_TYPES.include) {
+      const [, includeFilePath] = token.value.split(' ');
+
+      let includeConfig = '';
+      const fullPath = join(dirname(filePath), includeFilePath);
+
+      try {
+        const realPath = realpathSync(fullPath);
+
+        if (pathMemo.has(realPath)) {
+          throw new TokenError(
+            `You are trying to include ${realPath}. Fluent Bit does not allow the a file to be included twice in the same configuration`,
+            filePath,
+            token.line,
+            token.col
+          );
+        }
+
+        includeConfig = readFileSync(realPath, { encoding: 'utf-8' });
+        pathMemo.add(realPath);
+      } catch (e) {
+        if (e instanceof TokenError) {
+          throw e;
+        }
+        throw new TokenError(`Can not read file, loading from ${filePath}`, fullPath, token.line, token.col);
```

Comment:
Ah so this is where we'd handle invalid encoding, etc. I think

Reply:
we could include the original message from the error if this will make it more ergonomic.

Reply:
https://github.com/calyptia/fluent-bit-config-parser/issues/19